### PR TITLE
Database Authentication

### PR DESCRIPTION
**Pull request by danielpetisme**
Reproduce the DB+JWT JHipster authentication. #12 

Done
1. The user send its credentials to an open endpoint. 
2. Quarkus has no programatic way to invoke an `authenticate` method so there is a dedicated `AuthenticationService` which does  the Spring `UserDetailsService` and `AuthenticationManagerBuilder` job (ie loading a domain user with entities from a DB and checking BCrypted hasshed passwords
3. Once authenticated, the application forges a JWT. Quarkus is more rigourous on the security algo and thus RSA keys are mandatories (where JHipster can work with a base64 secret).
The app embeds a key pair (they are coming from Quarkus JWT guide).
Finally the app can return a JWT to the frontend for the next requests.
4. The app is configured with `quarkus-smallrye-jwt` which automatically intercept the request, validate the token, load the user details into the context and apply the endpoint authorization (versus the manual `JWTFilter` present in Spring JHipster)

Among the way I found couple of issues.
- the `NONSTRICT_READ_WRITE ` caching option works nice when I run the app but fails during the test (https://github.com/quarkusio/quarkus/issues/6492).
Meanwhile I switch to `READ_ONLY` just to move on and I track an issue in the current repo
- Quarkus as an injection issue with JUnit which makes impossible to inject properly a repository. Quarkus's team is working hard on this limit.
Meanwhile I manually look for a bean in CDI like:`UserRepository userRepository = CDI.current().select(UserRepository.class).get();`, it's not ideal but again it permits to move on.
- There is no version of Zalando problems for JAX-RS and the default JAX-RS exception handing is way too limited. IMHO, Integrating Zalando is not a big issue but this PR was already huge enough.

Bottom line, I have to polish some part but I succeed to authenticate via the frontend againt a Quarkus Server, and that's cool!!

### CODE REVIEW

**Comment by danielpetisme** [comment]
@amanganiello90 I would be more than happy if you could have a look to the PR, please.

**Comment by amanganiello90** [comment]
> 
> 
> @amanganiello90 I would be more than happy if you could have a look to the PR, please.

No problem, I will check it :)

**Comment by amanganiello90** [comment]
The problem is the prettier.. I can do a PR to add it. After, before pushing, you have to execute on the generator main dir:

> npm run lint-fix

Otherwise we can add a commit hook to run automatically it.

**Comment by danielpetisme** [comment]
I'm always getting trouble with prettier...  If you have a solution it would be great!

**Comment by amanganiello90** [comment]
> 
> 
> I'm always getting trouble with prettier... If you have a solution it would be great!

Yes, I have already added a prettier with #19 .. Now all is formatted and an hook check is run before git commit. If it fails, execute the lint fix command:

> npm run lint:fix

The previous correct the formatting errors, but if you have violated some eslint rules (such as variable unused or undeclared), you have to fix it manually (lint tells you where lines there are errors).

**Comment by danielpetisme** [comment]
@amanganiello90 the feature is ready to be reviewed if you have time to have a look.
Thanks

**Comment by danielpetisme** [comment]
? Not sure to understand what you.mean, sorry.

**Comment by amanganiello90** [comment]
> 
> 
> ? Not sure to understand what you.mean, sorry.

The idea is to not add templates in blueprint that are the same of template used in generator-jhipster, but inherit it. So we don't have maintenance and check it in case of a version upgrade of jhipster.
For example your [LoginVM](https://github.com/jhipster/jhipster-quarkus/blob/16d1895b7e742960dd84b2f77cab202ea69a9a71/generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs) seems to be equal to [JHipster LoginVM template](https://github.com/jhipster/generator-jhipster/blob/v6.6.0/generators/server/templates/src/main/java/package/web/rest/vm/LoginVM.java.ejs).
So in that case, you can do the following in your server/files.js:

```
const path = require('path');
const constants = require('generator-jhipster/generators/generator-constants');
const JHIPSTER_CLI_PATH = require.resolve('generator-jhipster/cli/jhipster.js');
const SERVER_MAIN_SRC_DIR = constants.SERVER_MAIN_SRC_DIR;
const JHIPSTER_RELATIVE_TEMPLATE_DIR = path.join(JHIPSTER_CLI_PATH, '..','..', 'generators', 'server', 'templates', SERVER_MAIN_SRC_DIR);

....
  serverJavaWeb: [
        {
            path: SERVER_MAIN_SRC_DIR,
            templates: [
                {
                    file: `${JHIPSTER_RELATIVE_TEMPLATE_DIR}package/web/rest/LoginVM.java.ejs`
                    renameTo: generator => `${generator.javaDir}web/rest/vm/LoginVM.java`
                },
....


```

File: generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<%#
```
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs]: Sorry we use flyway and not jhipster liquibase, so it is new.

File: generators/server/templates/quarkus/src/main/resources/META-INF/resources/privateKey.pem [hunk 2]
```diff
@@ -0,0 +1,28 @@
+-----BEGIN PRIVATE KEY-----
```
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/resources/META-INF/resources/privateKey.pem]: Sorry, Jhipster does not use already these certificates?

File: generators/server/templates/quarkus/src/main/resources/META-INF/resources/publicKey.pem [hunk 3]
```diff
@@ -0,0 +1,9 @@
+-----BEGIN PUBLIC KEY-----
```
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/resources/META-INF/resources/publicKey.pem]: Sorry, Jhipster does not use already these certificates?

[the remark this answers is not in the document]
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/main/resources/META-INF/resources/publicKey.pem]: Noppe... the JWT signing/validation is made by a secret not a key pair
https://github.com/jhipster/jhipster-sample-app/blob/master/src/main/resources/config/application-prod.yml#L122
Eventhough there is a property to provide a key by properties is not generated.

File: generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs]: 👍

File: generators/server/templates/quarkus/src/main/resources/db/migration/V1_00000000000000__Initial_schema.sql [hunk 4]
```diff
@@ -0,0 +1,71 @@
+DROP TABLE IF EXISTS jhi_user_authority;
```
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/resources/db/migration/V1_00000000000000__Initial_schema.sql]: Yes.. sorry , I have alredy "auto answered" on this: https://github.com/jhipster/jhipster-quarkus/pull/16#discussion_r367293026

File: generators/server/templates/quarkus/src/main/resources/META-INF/resources/publicKey.pem (hunk 3 above)
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/resources/META-INF/resources/publicKey.pem]: Ok, It was only an example to consider this for all that we can inherit: https://github.com/jhipster/jhipster-quarkus/pull/16#issuecomment-575046113

File: generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielpetisme) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs]: It's not the same template.
JHipster use getter and setter where Quarkus is pushing to make the the fields public.

From a functionnal stand point it's similar but it's part of the Quarkus mind switch.

**Comment by danielpetisme** [comment]
I made couple of comments. Feel free to close/resolve the conversations if you think I answered to the point

File: generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanganiello90) [reply generators/server/templates/quarkus/src/main/java/package/web/rest/vm/LoginVM.java.ejs]: Yes, you are right for the security topic... my advice was to maximize the reusability of jhipster templates and code, in order to avoid possible error code and to create new tests or coverage. Obviously where is possible to reuse code :)

**Comment by amanganiello90** [comment]
> 
> 
> I made couple of comments. Feel free to close/resolve the conversations if you think I answered to the point

Thanks @danielpetisme , it was already all ok, my discussion was only to improve or getting advices... that is the sense of a review, the good comparison to grow professionally :)

**Comment by danielpetisme** [comment]
👍 
Sorry, personal habit I wait for the reviewer to close the changes request just to be sure all the points are covered.
I had cases where I prematurely close a discussion 😢

**Comment by amanganiello90** [comment]
> 
> 
> 👍
> Sorry, personal habit I wait for the reviewer to close the changes request just to be sure all the points are covered.
> I had cases where I prematurely close a discussion 😢

Yes, all point are ok. You are a gentleman, don't warry :)
